### PR TITLE
#6433: fixes naive sign then encrypt in S2S

### LINKED ISSUE #6433: Naive sign then encrypt as we do in in S2S EncrpytionService is not secure
See https://github.com/hzi-braunschweig/SORMAS-Security/issues/20

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #6433

### CODE REVIEW

File: sormas-api/pom.xml [hunk 1]
```diff
@@ -11,6 +11,16 @@
 
 	<build>
 		<finalName>${project.artifactId}</finalName>
+        <plugins>
```
Comment (JonasCir) [inline sormas-api/pom.xml:14]: remove this

File: sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java [hunk 2]
```diff
@@ -41,17 +51,42 @@
...
+		}
+
+		if (!plaintext.getReceiverId().equals(certificateConfig.getOwnId())) {
+			LOGGER.error("Wrong receiver: Expected {}, was {}", certificateConfig.getOwnId(), plaintext.getReceiverId());
+			throw new SecurityException("We are not the intended receiver of the message!");
+		}
+
+		if (!plaintext.getSenderId().equals(certificateConfig.getOtherId())) {
+			LOGGER.error("Wrong sender: Expected {}, was {}", certificateConfig.getOtherId(), plaintext.getSenderId());
+			throw new SecurityException("The send of the message is not correct!");
```
Comment (JonasCir) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java:85]: typo

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java [hunk 3]
```diff
@@ -144,38 +145,38 @@ private X509Certificate getOtherCertificate(String otherId)
...
+			this.otherId = otherId;
+
+			KeyStore keystore = getKeystore();
+
+			this.ownCertificate = loadOwnCertificate();
+			this.ownPrivateKey = (PrivateKey) keystore.getKey(ownId, sormasToSormasConfig.getKeystorePass().toCharArray());
+			this.otherCertificate = loadOtherCertificate(otherId);
+
+			if (this.otherCertificate == null) {
+				throw SormasToSormasException.fromStringProperty(Strings.errorSormasToSormasCertNotGenerated);
```
Comment (JonasCir) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java:165]: logging missing, log success as well

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java [hunk 4]
```diff
@@ -144,38 +145,38 @@ private X509Certificate getOtherCertificate(String otherId)
...
 
 	@Override
 	public SormasToSormasEncryptedDataDto signAndEncrypt(Object entities, String recipientId) throws SormasToSormasException {
 		LOGGER.info("Sign and encrypt data for {}", recipientId);
 		try {
-			byte[] encryptedData = cipher(Mode.ENCRYPTION, objectMapper.writeValueAsBytes(entities), recipientId);
-			return new SormasToSormasEncryptedDataDto(configFacadeEjb.getS2SConfig().getId(), encryptedData);
+			final String ownId = configFacadeEjb.getS2SConfig().getId();
+			CmsPlaintext plaintext = new CmsPlaintext(ownId, recipientId, entities);
+			S2SCertificateConfig config = new S2SCertificateConfig(plaintext.getReceiverId());
```
Comment (JonasCir) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java:176]: use recipientId

File: sormas-api/pom.xml (hunk 1 above)
Comment (JonasCir) [reply sormas-api/pom.xml:14]: fixed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java (hunk 2 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java:85]: fixed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java (hunk 3 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java:165]: done

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java (hunk 4 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/sormastosormas/SormasToSormasEncryptionFacadeEjb.java:176]: done

File: sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java [hunk 5]
```diff
@@ -9,9 +10,16 @@
...
 import java.security.cert.X509Certificate;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import de.symeda.sormas.api.sormastosormas.SormasToSormasException;
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java:20]: `SormasToSormasException` and `SormasToSormasValidationException` don't seem to be used

**Review by vidi42 — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/java/de/symeda/sormas/api/sormastosormas/SormasToSormasEncryptionFacade.java [hunk 6]
```diff
@@ -1,6 +1,7 @@
 package de.symeda.sormas.api.sormastosormas;
 
 import java.io.IOException;
+import java.io.Serializable;
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-api/src/main/java/de/symeda/sormas/api/sormastosormas/SormasToSormasEncryptionFacade.java:4]: fixed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsPlaintext.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
+package de.symeda.sormas.backend.crypt;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.Serializable;
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsPlaintext.java:10]: fixed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java (hunk 5 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/crypt/CmsReader.java:20]: fixed

**Review by vidi42 — APPROVED** [review]

**Pull request merged**